File: leetcode/golang/238.除自身以外数组的乘积.go
func productExceptSelf(nums []int) []int {
    n := len(nums);
    if n == 1{
        return nums;
    }
    ret := make([]int, n);
    ret[0] = 1;
    for i, _ := range ret {
        if i == 0 {
            continue;
        }
        ret[i] = ret[i - 1] * nums[i - 1];
    }
    j := n - 2;
    t := 1;
    for j >= 0 {
        t *= nums[j + 1]
        ret[j] *= t;
        j -= 1;
    }
    return ret;
}

File: leetcode/python/238.除自身以外数组的乘积.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        if n == 1: return nums

        ret = [1 for _ in range(n)]
        for i in range(1, n):
            ret[i] = ret[i - 1] * nums[i - 1]

        j = n - 2
        tmp = 1
        while j >= 0:
            tmp *= nums[j + 1]
            ret[j] *= tmp
            j -= 1
        
        return ret